### PR TITLE
[doc] Add documentation to run tests and coverage

### PR DESCRIPTION
**Pull request by valeriocos**
This code enhances the README.md by adding the information to execute the tests. Furthermore details about how to use the coverage package are also included.

Signed-off-by: Valerio Cosentino <valcos@bitergia.com>

### CODE REVIEW

**Comment by valeriocos** [comment]
@snack0verflow, please try to follow the steps included in this PR, thanks

**Comment by valeriocos** [comment]
@snack0verflow did you have time to check this PR? If the steps are clear, we can merge the PR. Thanks

**Comment by snack0verflow** [comment]
@valeriocos sorry I got busy with some college assignments. Will I be able to run the tests from the ELK repository that I cloned for https://github.com/chaoss/grimoirelab-sirmordred/blob/master/README.md#setting-up-a-pycharm-dev-environment ? Should I also use the same pycharm project to raise PRs for coverage improvements?

**Comment by vchrombie** [comment]
Hi @valeriocos 

I just tried the things which you have mentioned in the README. 

You can view the results in this [gist](https://gist.github.com/vchrombie/135da00879e5f7c554a3e38de6c4234f). I see a lot of errors in the [log](https://gist.github.com/vchrombie/135da00879e5f7c554a3e38de6c4234f#file-run_tests_log) but the [results](https://gist.github.com/vchrombie/135da00879e5f7c554a3e38de6c4234f#file-run_tests_report) are smooth I guess. I can see the coverage report of the enriched files from [L622](https://gist.github.com/vchrombie/135da00879e5f7c554a3e38de6c4234f#file-run_tests_report-L622) to [L667](https://gist.github.com/vchrombie/135da00879e5f7c554a3e38de6c4234f#file-run_tests_report-L667) in the [file-run_tests_report](https://gist.github.com/vchrombie/135da00879e5f7c554a3e38de6c4234f#file-run_tests_report).

Can you please check them once?

---

Also, one more thing. The last point in the [README.md#enriched-data](https://github.com/chaoss/grimoirelab-elk/blob/master/README.md#enriched-data) is Twitter.

I think there is a mistake.
```
Telegram: each document corresponds to a message.
Telegram: each document corresponds to a tweet.
```

**Comment by valeriocos** [comment]
> You can view the results in this gist. I see a ...

The error is due to a recent version of MariaDb (more info at https://github.com/chaoss/grimoirelab-sortinghat/issues/121). Please try to use MariaDb 10.0 (https://github.com/chaoss/grimoirelab-sortinghat/issues/121#issuecomment-413505988)

> I think there is a mistake.

good catch @vchrombie , can you submit a PR to fix the mistake?

File: README.md [hunk 1]
```diff
@@ -112,3 +112,38 @@ profile information (i.e., identity fields) or (iv) to the project listed in the
...
+- populate the database `test_projects` with the SQL file [test_projects.sql](https://github.com/chaoss/grimoirelab-elk/blob/master/tests/test_projects.sql) (e.g., `mysql -u root test_projects < tests/test_projects.sql`)
+
+The full battery of tests can be executed with [run_tests.py](https://github.com/chaoss/grimoirelab-elk/blob/master/tests/run_tests.py). However, it is also possible to execute
+a sub-set of tests by running the single test files (`test_*` files in the [tests folder](https://github.com/chaoss/grimoirelab-elk/tree/master/tests))
+
+The tests can be run in combination with the Python package `coverage`. The steps below show how to do it:
+```buildoutcfg
+$ pip3 install coveralls
+$ cd <path-to-ELK>/tests
+$ coverage run run_tests.py --source=grimoire_elk
```
Comment (vchrombie) [inline README.md]: The command can be updated to 
```
python3 -m coverage run run_tests.py --source=grimoire_elk
```

File: README.md [hunk 2]
```diff
@@ -112,3 +112,38 @@ profile information (i.e., identity fields) or (iv) to the project listed in the
...
+```buildoutcfg
+$ pip3 install coveralls
+$ cd <path-to-ELK>/tests
+$ coverage run run_tests.py --source=grimoire_elk 
+```
+
+Coverage will generate a file `.coverage` in the tests folder, which can be inspected with the following command:
+```buildoutcfg
+cd <path-to-ELK>/tests
+coverage report -m
```
Comment (vchrombie) [inline README.md]: Same as the above comment.
```
python3 -m coverage report -m
```

**Comment by valeriocos** [comment]
@vchrombie thank you for the feedback, can you have a final look before merging the PR?

**Comment by vchrombie** [comment]
> can you have a final look before merging the PR?

The instructions look perfectly fine. I think we can merge it.

> If you have PyCharm set up, it would be great if you can enhance this new section with some details about how to run the tests from PyCharm. WDYT?

Sure, I would do that. :slightly_smiling_face:

**Comment by valeriocos** [comment]
thanks @vchrombie

**Comment by snack0verflow** [comment]
@valeriocos @vchrombie  Some steps here are not clear to me. How do I perform these steps:

> - create the databases test_sh and test_projects in your MySQL instance (e.g., mysql -u root -e "create database test_sh")
> - populate the database test_projects with the SQL file test_projects.sql (e.g., mysql -u root test_projects < tests/test_projects.sql)

I am using the docker-compose https://github.com/chaoss/grimoirelab-sirmordred#source-code-and-docker
When I type `mysql -u root -e "create database test_sh"` in a terminal, it says mysql is not installed.

**Comment by vchrombie** [comment]
Hi @snack0verflow 

> When I type `mysql -u root -e "create database test_sh"` in a terminal, it says mysql is not installed.

Ok. I have a few pointers. 

1. Make sure you have the correct version of the software installed. The suggested versions are MySQL/MariaDB database (5.7.24/10.0), [reference](https://github.com/chaoss/grimoirelab-sirmordred#getting-started). I would suggest you to install MariaDB 10.0, just to avoid any other errors, [reference](https://github.com/chaoss/grimoirelab-elk/pull/784#issuecomment-588888815).

    You can check the current version using
    ```
    mysql -V
    ```

2. If you are having any password for `root`, good to add `-p` flag to the command.

Can you please send me the error message so that I can look into it more?

---

> I am using the docker-compose https://github.com/chaoss/grimoirelab-sirmordred#source-code-and-docker

I am not sure about the docker configuration here. I installed MySQL and elasticsearch which worked perfectly fine. You can do one thing, use elasticsearch via docker and install MySQL or MariaDB. I hope that works. :slightly_smiling_face:

**Comment by valeriocos** [comment]
> I am not sure about the docker configuration here. I installed MySQL and elasticsearch which worked perfectly fine. You can do one thing, use elasticsearch via docker and install MySQL or MariaDB. I hope that works...

The solution suggested by @vchrombie is good

If you want to use the docker-compose solution, you need to run the mysql commands from the container (docker exec -it <container-name> bash)

**Comment by snack0verflow** [comment]
@valeriocos Thank you! As someone who has never used docker before I think the documentation is seriously lacking. Should I raise a PR suggesting some changes to the readme of ELK and sirmordred repos?

**Comment by valeriocos** [comment]
you're welcome @snack0verflow ! Yes, indeed. Can we start with the PR for ELK and then we will move to sirmordred?

**Comment by snack0verflow** [comment]
Hi @valeriocos! I am getting a 401 unauthenticated request while running these tests:
```
Traceback (most recent call last):
  File "../grimoire_elk/utils.py", line 268, in get_elastic
    analyzers=analyzers, aliases=es_aliases)
  File "../grimoire_elk/elastic.py", line 62, in __init__
    self.major = self.check_instance(url, insecure)
  File "../grimoire_elk/elastic.py", line 120, in check_instance
    raise ElasticError(cause=msg)
grimoire_elk.errors.ElasticError: Got 401 from url https://localhost:9200
```
I am using the secured version of elasticsearch as specified in [this readme](https://github.com/chaoss/grimoirelab-sirmordred#source-code-and-docker).

Also, I had to install `statsmodels` module which was not mentioned in the `requirements_tests.txt`.

**Comment by valeriocos** [comment]
Did you update the tests.conf with https://admin:admin@localhost:9200? This is needed if you are accessing a secured elasticsearch.

The statsmodels is listed as a requirement (https://github.com/chaoss/grimoirelab-elk/blob/6f449a69d8730d88712bb6332227808a6bdadd4a/requirements.txt#L9). In the requirements_tests we generally add the packages used for testings (e.g., httpretty).

**Comment by snack0verflow** [comment]
> Did you update the tests.conf with https://admin:admin@localhost:9200? This is needed if you are accessing a secured elasticsearch.

@valeriocos I had set tests.conf to https://localhost:9200, after setting it to https://admin:admin@localhost:9200, every single test passed. Thanks :) should I submit a PR for some changes to this doc?

>  In the requirements_tests we generally add the packages used for testings (e.g., httpretty).

My bad, I must have missed it during installation.
Looks like finally I can work on increasing the coverage.

**Comment by valeriocos** [comment]
> ...very single test passed. Thanks :) should I submit a PR for some changes to this doc?

You're welcome! Yes, it's a good idead to submit a PR :). Maybe, you can modify the part below:
```
Then you need to:

update the file tests.conf in case your ElasticSearch instance isn't available at http://localhost:9200.
```

Thanks!

**Pull request merged**